drafts/records.md

```diff
@@ -0,0 +1,398 @@
+
+A `record` body may contain property hooks, methods, and use traits (so long as they do not conflict with `record`
+rules).
+Regular properties may also be defined, but they are immutable by default and are no different from `const`.
+
+Static properties and methods are forbidden in a `record` (this includes
+`const`, a regular property may be used instead). Attempting to define
+static properties, methods, constants results in a compilation error.
+
+``` php
+namespace Paint;
+
+record Pigment(int $red, int $yellow, int $blue) {
+  public function mix(Pigment $other, float $amount): Pigment {
+    return $this->with(
+      red: $this->red * (1 - $amount) + $other->red * $amount,
+      yellow: $this->yellow * (1 - $amount) + $other->yellow * $amount,
+      blue: $this->blue * (1 - $amount) + $other->blue * $amount
+    );
+  }
+}
+
+record StockPaint(Pigment $color, float $volume);
+
+record PaintBucket(StockPaint ...$constituents) {
+  public function mixIn(StockPaint $paint): PaintBucket {
+    return $this->with(...$this->constituents, $paint);
+  }
+  
+  public function color(): Pigment {
+    return array_reduce($this->constituents, fn($color, $paint) => $color->mix($paint->color, $paint->volume), Pigment(0, 0, 0));
+  }
+}
+```
+
+#### Usage
+
+A `record` may be used as a `readonly class`,
+as the behavior of it is very similar with no key differences to assist in migration from `readonly class`.
+
+#### Optional parameters and default values
+
+A `record` can also be defined with optional parameters that are set if omitted during instantiation.
+
+``` php
+record Rectangle(int $x, int $y = 10);
+var_dump(Rectangle(10)); // output a record with x: 10 and y: 10
+```
+
+#### Auto-generated `with` method
+
+To enhance the usability of records, the RFC proposes automatically generating a `with` method for each record.
+This method allows for partial updates of properties, creating a new instance of the record with the specified
+properties updated.
+
+The auto-generated `with` method accepts only named arguments defined in the constructor.
+No other property names can be used, and it returns a record object with the given values.
+
+``` php
+$point1 = Point(3, 4);
+$point2 = $point1->with(x: 5);
+$point3 = $point1->with(null, 10); // must use named arguments
+
+echo $point1->x; // Outputs: 3
+echo $point2->x; // Outputs: 5
+```
+
+A developer may define their own `with` method if they so choose,
+and reference the generated `with` method using `parent::with()`.
+This allows a developer to define policies or constraints on how data is updated.
+
+``` php
+record Planet(string $name, int $population) {
+  public function with(int $population) {
+    return parent::with(population: $population);
+  }
+}
+$pluto = Planet("Pluto", 0);
+// we made it!
+$pluto = $pluto->with(population: 1);
+// and then we changed the name
+$mickey = $pluto->with(name: "Mickey"); // no named argument for population error
+```
+
+#### Constructors
```

Comment:
Maybe _"**Explicit Constructors**?"_

That could mean that _"Inline Constructors"_ could be named _"Implicit Constructors,"_ maybe?